### PR TITLE
Snowball tracing prototype

### PR DESCRIPTION
**Pull request by tbg**
opentracing-go's API is still under some flux, so here is an experiment of
"snowball propagation" for a single request in the SQL shell.

By snowball propagation I mean that the segments of the trace are serialized
and sent back to the client (sql gateway). This isn't practical for distributed
tracing generally, but for intentionally tracing a single element it seemed
easy enough to set up.

For those of us not too familiar with distributed tracing, Dapper at Google
simply wrote all trace spans to local disk where they happened, and a separate
system (Dapper daemon) would come along, collect trace spans and write them to
a database for visualization and analysis. Of course, we're a database and so
the obvious suggestion is that selected traces (chosen probabilistically or by
some criterion) are persisted in a SQL table where they are created (i.e.
different nodes each log the part of the request that is traced on them). I
didn't want to get into setting up that infrastructure just yet, though.

I was curious whether in-session viewing of traces could be useful at all. The
layout here is a very first iteration, and the goal would be to give
us/admins/devs an easy way of gauging how the distributed aspects of a query
pan out (similar to how `EXPLAIN (DEBUG)` tries to explain SQLs strategy for
fetching the result). In any case, the current format is almost likely
confusing and feedback appreciated. Maybe not everything ought to be shown (and
much is omitted here). For some MySQL-related tracing output, see [1].

```
root@:15432> explain (trace) select * from x.y;
+-----+-----+-----------+------------------------------------+--------+------------------+--------+--------+
| ... | Pos | Operation |               Event                | RowIdx |       Key        | Value  | Output |
+-----+-----+-----------+------------------------------------+--------+------------------+--------+--------+
| ... |   0 | "sql"     | "sending batch"                    |      0 | NULL             | ""     | NULL   |
| ... |   1 | "sql"     | "meta descriptor lookup"           |      0 | NULL             | ""     | NULL   |
| ... |   2 | "sql"     | "sending RPC to [/Table/50, /Max)" |      0 | NULL             | ""     | NULL   |
| ... |   3 | "sql"     | "sending to [::]:26257"            |      0 | NULL             | ""     | NULL   |
| ... |   0 | "node"    | "node 1"                           |      0 | "/y/primary/1"   | "NULL" | NULL   |
| ... |   1 | "node"    | "read has no clock uncertainty"    |      0 | "/y/primary/1"   | "NULL" | NULL   |
| ... |   2 | "node"    | "executing 1 requests"             |      0 | "/y/primary/1"   | "NULL" | NULL   |
| ... |   3 | "node"    | "read-only path"                   |      0 | "/y/primary/1"   | "NULL" | NULL   |
| ... |   4 | "node"    | "command queue"                    |      0 | "/y/primary/1"   | "NULL" | NULL   |
| ... |   5 | "node"    | "request leader lease"             |      0 | "/y/primary/1"   | "NULL" | NULL   |
| ... |   6 | ""        | ""                                 |      0 | "/y/primary/1/s" | "'a'"  | true   |
| ... |   7 | ""        | ""                                 |      1 | "/y/primary/2"   | "NULL" | NULL   |
| ... |   8 | ""        | ""                                 |      1 | "/y/primary/2/s" | "'b'"  | true   |
| ... |   9 | ""        | ""                                 |      2 | "/y/primary/3"   | "NULL" | NULL   |
| ... |  10 | ""        | ""                                 |      2 | "/y/primary/3/s" | "'c'"  | true   |
| ... |   4 | "sql"     | "tracing completed"                |      0 | NULL             | ""     | NULL   |
+-yp--+-----+-----------+------------------------------------+--------+------------------+--------+--------+
```

Benchmarks below. The actual tracing has overhead, but it's still disabled
(trace encoding is still pretty horrible). Only a single run since allocs are
the interesting datapoint.

```
name                    old time/op    new time/op    delta
Insert1_Cockroach-32       581µs ± 0%     585µs ± 0%   ~             (p=1.000 n=1+1)
Insert10_Cockroach-32      992µs ± 0%     999µs ± 0%   ~             (p=1.000 n=1+1)
Insert100_Cockroach-32    4.43ms ± 0%    4.43ms ± 0%   ~             (p=1.000 n=1+1)
Update1_Cockroach-32       876µs ± 0%     888µs ± 0%   ~             (p=1.000 n=1+1)
Update10_Cockroach-32     1.73ms ± 0%    1.77ms ± 0%   ~             (p=1.000 n=1+1)
Update100_Cockroach-32    8.49ms ± 0%    8.55ms ± 0%   ~             (p=1.000 n=1+1)
Delete1_Cockroach-32      1.11ms ± 0%    1.11ms ± 0%   ~             (p=1.000 n=1+1)
Delete10_Cockroach-32     2.88ms ± 0%    2.84ms ± 0%   ~             (p=1.000 n=1+1)
Delete100_Cockroach-32    19.0ms ± 0%    19.2ms ± 0%   ~             (p=1.000 n=1+1)
Scan1_Cockroach-32         303µs ± 0%     290µs ± 0%   ~             (p=1.000 n=1+1)
Scan10_Cockroach-32        343µs ± 0%     347µs ± 0%   ~             (p=1.000 n=1+1)
Scan100_Cockroach-32       768µs ± 0%     762µs ± 0%   ~             (p=1.000 n=1+1)

name                    old alloc/op   new alloc/op   delta
Insert1_Cockroach-32      24.0kB ± 0%    24.0kB ± 0%   ~             (p=1.000 n=1+1)
Insert10_Cockroach-32     74.9kB ± 0%    74.9kB ± 0%   ~             (p=1.000 n=1+1)
Insert100_Cockroach-32     586kB ± 0%     586kB ± 0%   ~             (p=1.000 n=1+1)
Update1_Cockroach-32      34.6kB ± 0%    34.6kB ± 0%   ~             (p=1.000 n=1+1)
Update10_Cockroach-32      124kB ± 0%     124kB ± 0%   ~             (p=1.000 n=1+1)
Update100_Cockroach-32     965kB ± 0%     965kB ± 0%   ~             (p=1.000 n=1+1)
Delete1_Cockroach-32      35.5kB ± 0%    35.5kB ± 0%   ~             (p=1.000 n=1+1)
Delete10_Cockroach-32      137kB ± 0%     136kB ± 0%   ~             (p=1.000 n=1+1)
Delete100_Cockroach-32    1.10MB ± 0%    1.10MB ± 0%   ~             (p=1.000 n=1+1)
Scan1_Cockroach-32        9.95kB ± 0%    9.92kB ± 0%   ~             (p=1.000 n=1+1)
Scan10_Cockroach-32       15.6kB ± 0%    15.6kB ± 0%   ~             (p=1.000 n=1+1)
Scan100_Cockroach-32      61.0kB ± 0%    60.9kB ± 0%   ~             (p=1.000 n=1+1)

name                    old allocs/op  new allocs/op  delta
Insert1_Cockroach-32         325 ± 0%       325 ± 0%   ~     (all samples are equal)
Insert10_Cockroach-32        810 ± 0%       810 ± 0%   ~     (all samples are equal)
Insert100_Cockroach-32     5.39k ± 0%     5.39k ± 0%   ~             (p=1.000 n=1+1)
Update1_Cockroach-32         564 ± 0%       564 ± 0%   ~     (all samples are equal)
Update10_Cockroach-32      1.40k ± 0%     1.40k ± 0%   ~     (all samples are equal)
Update100_Cockroach-32     9.12k ± 0%     9.13k ± 0%   ~             (p=1.000 n=1+1)
Delete1_Cockroach-32         572 ± 0%       572 ± 0%   ~     (all samples are equal)
Delete10_Cockroach-32      1.76k ± 0%     1.75k ± 0%   ~             (p=1.000 n=1+1)
Delete100_Cockroach-32     12.9k ± 0%     12.9k ± 0%   ~             (p=1.000 n=1+1)
Scan1_Cockroach-32           182 ± 0%       181 ± 0%   ~             (p=1.000 n=1+1)
Scan10_Cockroach-32          284 ± 0%       284 ± 0%   ~     (all samples are equal)
Scan100_Cockroach-32       1.20k ± 0%     1.20k ± 0%   ~     (all samples are equal)
```

[1] http://jorgenloland.blogspot.com/2011/10/optimizer-tracing-query-execution-plan.html

[Review on Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/4178)

### CODE REVIEW

**Comment by tamird** [comment]
Exciting work.

---

Reviewed 24 of 24 files at r1.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[roachpb/api.proto, line 648 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4178#-K9qpcCsxk5BVnsnUxE1:-K9qpcCtZe95Sh5vrEPi:359409460) ([raw file](https://github.com/cockroachdb/cockroach/blob/e576d97a2b7e6fbdf69d54e606a22a4d4a303a84/roachpb/api.proto#L648)):*
can you add a comment?

---

*[server/node.go, line 553 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4178#-K9qschIhD8RBBxrtNU1:-K9qschJaJvE0FWRcDak:-286864956) ([raw file](https://github.com/cockroachdb/cockroach/blob/e576d97a2b7e6fbdf69d54e606a22a4d4a303a84/server/node.go#L553)):*
nit; FormatInt(int64(..))

---

*[sql/trace.go, line 52 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4178#-K9qvc3O9cggMfXuQtZw:-K9qvc3O9cggMfXuQtZx:-1512523823) ([raw file](https://github.com/cockroachdb/cockroach/blob/e576d97a2b7e6fbdf69d54e606a22a4d4a303a84/sql/trace.go#L52)):*
this method is quite hard to follow, can you sprinkle some comments?

---

*[util/tracing/span.proto, line 17 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4178#-K9qpPNTEdYIt-zvACLi:-K9qpPNTEdYIt-zvACLj:-1554416409) ([raw file](https://github.com/cockroachdb/cockroach/blob/e576d97a2b7e6fbdf69d54e606a22a4d4a303a84/util/tracing/span.proto#L17)):*
might want to use proto3

---

*[util/tracing/span.proto, line 23 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4178#-K9qpKZzZAgi5drXaPpp:-K9qpK_-ppspprEvpCBD:-1329008392) ([raw file](https://github.com/cockroachdb/cockroach/blob/e576d97a2b7e6fbdf69d54e606a22a4d4a303a84/util/tracing/span.proto#L23)):*
can you add a comment for each field?

---

*[util/tracing/tracer.go, line 112 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4178#-K9qqf2lYQK7A150yBzM:-K9qqf2m_iBb34ksEhC2:677500536) ([raw file](https://github.com/cockroachdb/cockroach/blob/e576d97a2b7e6fbdf69d54e606a22a4d4a303a84/util/tracing/tracer.go#L112)):*
as a

---

*[util/tracing/tracer.go, line 114 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4178#-K9qqm8eRSINz1Ah5qy_:-K9qqm8eRSINz1Ah5qya:217597371) ([raw file](https://github.com/cockroachdb/cockroach/blob/e576d97a2b7e6fbdf69d54e606a22a4d4a303a84/util/tracing/tracer.go#L114)):*
are there any alternatives on the horizon?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4178#-:-K9qvriJl8501qsPeDej:617294089)*

**Comment by petermattis** [comment]
I think this is going to be very useful. Especially once we start sprinkling trace logs in the `sql` layer as well.

---

Review status: all files reviewed at latest revision, 12 unresolved discussions.

---

*[roachpb/api.proto, line 606 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4178#-K9rHRrYz7vd7HPyTUPw:-K9rHRrYz7vd7HPyTUPx:-827724547) ([raw file](https://github.com/cockroachdb/cockroach/blob/e576d97a2b7e6fbdf69d54e606a22a4d4a303a84/roachpb/api.proto#L606)):*
Extend this comment to note that this field is not set if tracing is not desired.

---

*[server/node.go, line 533 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4178#-K9rIc3ZHM2OeR19CAIk:-K9rIc3ZHM2OeR19CAIl:486193614) ([raw file](https://github.com/cockroachdb/cockroach/blob/e576d97a2b7e6fbdf69d54e606a22a4d4a303a84/server/node.go#L533)):*
It would be a much more efficient usage of Gob to encode a slice of spans instead of encoding them one by one. Given how you're using `EncodeRawSpan`, this doesn't look like too difficult of a change. Just gather up the `rawSpans` into a slice here and add a defer that encodes them in one shot into a single `BatchRequest.collected_spans` field. On the decoding side you would similarly decode a `[]*standardtracer.RawSpan`. The advantage of this approach is that Gob would only output the type info for `standardtracer.RawSpan` once for all of the spans instead of once per span. And Gob creates custom encoders for every type it sees which would also be amortized instead of repeated for every span. 

---

*[server/node.go, line 553 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4178#-K9qschIhD8RBBxrtNU1:-K9rHAVpDTqKljWbMmvg:-187160435) ([raw file](https://github.com/cockroachdb/cockroach/blob/e576d97a2b7e6fbdf69d54e606a22a4d4a303a84/server/node.go#L553)):*
The `opentracing` APIs really need to differ the formatting cost in cases where the span isn't active. You might suggest upstream that they add `sp.LogEventf`.

---

*[sql/trace.go, line 46 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4178#-K9rLZL5SjBJZ1DtJl9T:-K9rLZL5SjBJZ1DtJl9U:1013049266) ([raw file](https://github.com/cockroachdb/cockroach/blob/e576d97a2b7e6fbdf69d54e606a22a4d4a303a84/sql/trace.go#L46)):*
I'm not sure that the trace should include the `debug` columns. For a query which touches a large number of rows the debug info could obscure the trace info. Perhaps you can make those columns optional by supporting `EXPLAIN (trace,debug)`?

Rather than a `Timestamp` field, can we show the duration of the span? 

---

*[sql/trace.go, line 52 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4178#-K9qvc3O9cggMfXuQtZw:-K9rKUHw0v3GMJ7Hb1MW:1591933725) ([raw file](https://github.com/cockroachdb/cockroach/blob/e576d97a2b7e6fbdf69d54e606a22a4d4a303a84/sql/trace.go#L52)):*
I think `explainTraceNode` would be better structured like `sortNode`. `sortNode.Next` is pretty simple. If the sub-plan's values have not been sorted, then they are read into `valuesNode` and then sorted and then the `valuesNode` replaces `sortNode.plan`. Happy to walk through in more detail how this works on Monday.

---

*[sql/trace.go, line 60 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4178#-K9rL9BVWnQ4ibKZSWR-:-K9rL9BVWnQ4ibKZSWR0:584651532) ([raw file](https://github.com/cockroachdb/cockroach/blob/e576d97a2b7e6fbdf69d54e606a22a4d4a303a84/sql/trace.go#L60)):*
You should be checking `n.plan.PErr` here. 

---

*[util/tracing/span.proto, line 27 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4178#-K9rGcM4CAyY4kV5_9Gq:-K9rGcM4CAyY4kV5_9Gr:223550232) ([raw file](https://github.com/cockroachdb/cockroach/blob/e576d97a2b7e6fbdf69d54e606a22a4d4a303a84/util/tracing/span.proto#L27)):*
`s/context/context_id/g`. You'd want a custom name (i.e. `ContextID`).

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4178#-:-K9rMtbLOYPdnmKU1Rda:615944770)*

**Comment by tschottdorf** [comment]
Review status: all files reviewed at latest revision, 12 unresolved discussions.

---

*[server/node.go, line 533 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4178#-K9rIc3ZHM2OeR19CAIk:-K9whsYi66SAD1c6qslA:289895029) ([raw file](https://github.com/cockroachdb/cockroach/blob/e576d97a2b7e6fbdf69d54e606a22a4d4a303a84/server/node.go#L533)):*
yes, but there's actually only one Span at the moment (and no plans for a change). Besides, performance isn't really a concern in this path. Let's get to that when the code is less experimental.

---

*[server/node.go, line 553 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4178#-K9qschIhD8RBBxrtNU1:-K9wi6GUFDsDkMGdqQVN:-925921902) ([raw file](https://github.com/cockroachdb/cockroach/blob/e576d97a2b7e6fbdf69d54e606a22a4d4a303a84/server/node.go#L553)):*
Yes, that stuff is all still in flux but I've been pushing for lazy logging.

---

*[sql/trace.go, line 46 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4178#-K9rLZL5SjBJZ1DtJl9T:-K9wiJuY7saBmbx7C_mG:-42846990) ([raw file](https://github.com/cockroachdb/cockroach/blob/e576d97a2b7e6fbdf69d54e606a22a4d4a303a84/sql/trace.go#L46)):*
Did you see my comment about being able to use `EXPLAIN` in subselects?

`SELECT timestamp, event FROM (EXPLAIN (TRACE) ....)` would make this more flexible.

Besides, mostly an experiment here. I think I agree with you that the debug output generally isn't as useful.

---

*[sql/trace.go, line 52 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4178#-K9qvc3O9cggMfXuQtZw:-K9wuKv2ji6hgq22NPU6:1481288201) ([raw file](https://github.com/cockroachdb/cockroach/blob/e576d97a2b7e6fbdf69d54e606a22a4d4a303a84/sql/trace.go#L52)):*
`sortNode` pulls everything into memory. Sorting is somewhat exceptional in that that's mostly necessary. Tracing should actually use the `Next()` api and not just buffer everything, especially since each call to `Next()` can create new trace activity and we want to keep track of which call did what (at least in some form, for example based on the index in the result set).
Agree though that the code is spaghetti, let's chat on Monday (also with respect to the next comment below).

---

*[sql/trace.go, line 60 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4178#-K9rL9BVWnQ4ibKZSWR-:-K9wmnH-c2Fj7Vlof2wQ:-422479019) ([raw file](https://github.com/cockroachdb/cockroach/blob/e576d97a2b7e6fbdf69d54e606a22a4d4a303a84/sql/trace.go#L60)):*
I changed `PErr()` on this node to return nil, and added the check for PErr here. Unfortunately, this doesn't really help with INSERT because that actually does all the work before we even construct this node. There's some refactoring that needs to happen so that INSERT on a unique constraint violation doesn't bail when on `makePlan` but instead returns a node which then gives the error. That way we can get a trace out of a failed INSERT as well.
On the other hand, trace timestamps are distributed, so there aren't many monotonicity guarantees, which speaks in favor of pulling everything (maybe we can annotate calls to the underlying `Next()` on the root span of the trace instead). I'm currently wrapping everything in a sort node, but that really gets in the way of displaying durations, etc (though I've added them within the same span).

There are similar problems on the batch level. I would expect that somewhere along we lose the trace due to the error-or-response dichotomy. Can be fixed, but I want to focus on creating useful output for now.

---

*[util/tracing/span.proto, line 23 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4178#-K9qpKZzZAgi5drXaPpp:-K9wnTK4P1GNOyqWjOF8:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/e576d97a2b7e6fbdf69d54e606a22a4d4a303a84/util/tracing/span.proto#L23)):*
Done.

---

*[util/tracing/span.proto, line 27 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4178#-K9rGcM4CAyY4kV5_9Gq:-K9wiiJlZne9Ufsp-7GE:-515762701) ([raw file](https://github.com/cockroachdb/cockroach/blob/e576d97a2b7e6fbdf69d54e606a22a4d4a303a84/util/tracing/span.proto#L27)):*
This is the serialized span context, so I don't think we should rename it. I'm renaming this struct to make clearer that it holds an encoded representation of the Span's metadata.

---

*[util/tracing/tracer.go, line 114 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4178#-K9qqm8eRSINz1Ah5qy_:-K9wkG8XPJs9rUh6G5qe:543144207) ([raw file](https://github.com/cockroachdb/cockroach/blob/e576d97a2b7e6fbdf69d54e606a22a4d4a303a84/util/tracing/tracer.go#L114)):*
Yes, really the only reason for this is the convenience and waiting until the fallout at opentracing-go settles.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4178)*

**Comment by tamird** [comment]
Reviewed 12 of 12 files at r2.
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[sql/trace.go, line 63 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4178#-K9xBmoh_CBtyuGpAOey:-K9xBmoi_4VMjYmvawI8:900434546) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb0b76ca98c9eea430d2e485dd433dafd7e381d4/sql/trace.go#L63)):*
you should call this `pErr`

---

*[sql/trace.go, line 86 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4178#-K9xBzBtkk1Xh3Gf8wp7:-K9xBzBuSHadd86hpvQ3:-882051017) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb0b76ca98c9eea430d2e485dd433dafd7e381d4/sql/trace.go#L86)):*
call String directly?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4178)*

**Comment by tschottdorf** [comment]
Good and bad news: the good news is that this is rebased and working again. The bad news is that the benchmarks suffer a bit. I'll take another pass at improving this in the morning.

**Comment by tschottdorf** [comment]
Ok, ready for a good look. The `explainTraceNode` is still pretty rudimentary, but that seems like a good follow-up. We now use "real" tracers (though nothing records those traces except when snowballing), so things are getting more serious.
I need to find a good way to efficiently marshal the trace into RPCs; we use protos, so really we would want to mirror the internal fields and just copy things into our proto (as opposed to storing binary in proto fields as we do now). I will look into making that happen upstream; there's an agreement that the current state is bad and that using protos is a good idea, so that should be doable.

**Comment by petermattis** [comment]
LGTM. What are the updated benchmark numbers?

---

Review status: 1 of 32 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[client/txn.go, line 56 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4178#-KBK2D0ImsPR8yx1BXzY:-KBK2D0ImsPR8yx1BXzZ:1031671018) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d0c7971d9599f713165daf0f102053ef68a8995/client/txn.go#L56)):*
Does `ContextWithSpan` do the right thing of `ts.Trace == nil`?

---

*[kv/dist_sender.go, line 392 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4178#-KBK2YvzFeGb8VmhA6u_:-KBK2YvzFeGb8VmhA6ua:-867729810) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d0c7971d9599f713165daf0f102053ef68a8995/kv/dist_sender.go#L392)):*
I'm guessing we're suffering a bit on these unnecessary `fmt.Sprintf` when tracing is disabled. We need `trace.LogEventf`.

---

*[sql/trace.go, line 46 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4178#-K9rLZL5SjBJZ1DtJl9T:-KBK3fyLg4AwtH9UWDA2:1794053223) ([raw file](https://github.com/cockroachdb/cockroach/blob/e576d97a2b7e6fbdf69d54e606a22a4d4a303a84/sql/trace.go#L46)):*
Ping. Are you planning to keep the `debug` columns?

---

*[util/tracing/span.proto, line 28 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4178#-KBK4YIFbOqnGVeriIFS:-KBK4YIFbOqnGVeriIFT:-53136605) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d0c7971d9599f713165daf0f102053ef68a8995/util/tracing/span.proto#L28)):*
We usually mark proto fields as explicitly `optional`.

---

*[util/tracing/tracer.go, line 70 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4178#-KBK4vllJiAcmmuXPVue:-KBK4vllJiAcmmuXPVuf:1793958682) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d0c7971d9599f713165daf0f102053ef68a8995/util/tracing/tracer.go#L70)):*
Document why this isn't a regular function.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4178#-:-KBK6pEkViUMikUPg-hc:1962930185)*

**Comment by tamird** [comment]
Reviewed 11 of 30 files at r3.
Review status: 12 of 32 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[util/tracing/span.proto, line 28 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4178#-KBK4YIFbOqnGVeriIFS:-KBKB-OxhCa8X85InQWV:-1715037266) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d0c7971d9599f713165daf0f102053ef68a8995/util/tracing/span.proto#L28)):*
That's not a thing in proto3.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4178)*

**Comment by tschottdorf** [comment]
Review status: 12 of 32 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[client/txn.go, line 56 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4178#-KBK2D0ImsPR8yx1BXzY:-KBNyGxE9ZR6jwFgXqa3:-1649211630) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d0c7971d9599f713165daf0f102053ef68a8995/client/txn.go#L56)):*
yes, it'll set a `nil` into the context but that's fine (you'll get a `nil` back later).

---

*[kv/dist_sender.go, line 392 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4178#-KBK2YvzFeGb8VmhA6u_:-KBNyP1o8JQF9uClumWi:-1530771663) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d0c7971d9599f713165daf0f102053ef68a8995/kv/dist_sender.go#L392)):*
Agreed.

---

*[sql/trace.go, line 46 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4178#-K9rLZL5SjBJZ1DtJl9T:-KBNyjQGrneUSeG3hm22:190125125) ([raw file](https://github.com/cockroachdb/cockroach/blob/e576d97a2b7e6fbdf69d54e606a22a4d4a303a84/sql/trace.go#L46)):*
I was hoping to just get this in for the backend stuff and fix up `EXPLAIN (DEBUG)` (which in a lot of ways is lacking) separately.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4178)*

**Comment by petermattis** [comment]
Review status: 10 of 32 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[kv/dist_sender.go, line 392 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4178#-KBK2YvzFeGb8VmhA6u_:-KBO-n3hsXUgqzXMX0uD:-1155281966) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d0c7971d9599f713165daf0f102053ef68a8995/kv/dist_sender.go#L392)):*
Can we wrap `opentracing.Span`and implement `LogEventf` ourselves? Or is the lazy log formatting coming soon enough to opentracing that this won't matter?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4178)*

**Comment by tschottdorf** [comment]
Review status: 7 of 32 files reviewed at latest revision, 7 unresolved discussions.

---

*[kv/dist_sender.go, line 392 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4178#-KBK2YvzFeGb8VmhA6u_:-KBO0hZnAB1mLbmW8g8Z:853980630) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d0c7971d9599f713165daf0f102053ef68a8995/kv/dist_sender.go#L392)):*
Wrapping is generally a real hassle (and also not free). Upstream will have a solution for this at some point and I don't think this does more than tickle benchmark OCD (we can always wrap the worst offenders in a check whether the trace is actually active).

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4178)*
